### DISCUSSION TITLE
Using Pyright as feedback for searching efficiently through ML-determined type annotations

### DISCUSSION THREAD

**Discussion by MarkBekooy** [microsoft/pyright#5868] [Q&A]
Hi, I have a question regarding using Pyright to get feedback on searching through combinations of possible machine learning predictions for type annotations.

Say I have the following very small and simple Python file:

```python
def multiply(a, b):
    return a * b
```

This has the following 3 type annotation slots:
Slot 1 = Parameter a
Slot 2 = Parameter b
Slot 3 = Function return type

I run a machine learning model to determine potential type annotations which returns the following array with predictions for each type slot.
```python
[
  [ # Slot 1
    ["int", 0.58],
    ["float", 0.41],
    ["bool", 0.05],
  ],
  [ # Slot 2
    ["float", 0.22],
    ["int", 0.20],
    ["bool", 0.09]
  ],
  [ # Slot 3
    ["str", 0.37],
    ["int", 0.32],
    ["float", 0.10],
  ]
];
```

I'd like to use the type-checking analysis of Pyright as feedback to gradually try combinations to see which are valid.
So I first add `"int"` to slot 1, run Pyright, which succeeds, then go to the next slot. In slot 2 I now fill in `"float"`, run Pyright, this also succeeds. Then for slot 3, I try `"str"` and Pyright fails, so I try `"int"` and Pyright fails, so I try `"float"` and Pyright succeeds. So I'll keep the combination of `["int", "float", "float"]` for the type slots. (The Pyright feedback can thus be used to be able to backtrack and try new combinations)

**Now, how can I get Pyright to constantly retrigger its type-checking analysis when I add a type annotation to my Python file?**
Should I create an LSP server that adds/changes the type annotation in the Python file and then the Pylance/Pyright LSP gets triggered automatically and I can use its diagnostics to determine whether the type annotation was successful?
Or is there a better way to do this?

I've looked into using the [@zzzen/pyright-internal](https://www.jsdelivr.com/package/npm/@zzzen/pyright-internal) npm package to use the internals of Pyright to retrigger the analysis, but possibly the LSP is a better idea? And otherwise, a suggestion for another way of accomplishing this would also be very much appreciated.

Thanks already!

**Comment by erictraut** [comment]
I think there are advantages and disadvantages to both approaches. 

The internal interface would give you more control, but this interface isn't documented or guaranteed to be stable. In fact, it changes quite frequently as pyright evolves. That means you're signing up for a large support burden. You'd need to build (and continually rebuild) the solution as pyright evolves.

By contrast, the LSP interface is limited to the functionality defined by the language server protocol, but it's a stable interface. The version of pyright that I publish each week on npm includes both a command-line (CLI) and a language server. If you go this route, you would technically be writing a "language server client" that talks to the pyright language server. Your client would send the language server LSP commands as though the user was manually editing the text within the file, and the pyright language server would re-analyze the file and respond with a new list of diagnostics.

I recommend starting with the LSP approach and see if it meets your needs. Based on your description above, I think it will.

**Reply by MarkBekooy** [reply]
Thank you kindly for the response @erictraut! I appreciate it and it has helped me out already.

I have some other questions related to retrieving this efficient feedback.

Looking through the pyright-internal modules, it seems like Pyright is generating an abstract syntax tree (AST) in memory for each parsed Python file. Now, when I manually make changes to the type annotations of my Python files in VS Code, I immediately get feedback since these files are being "watched". For this I suppose the internal AST also constantly and rapidly gets updated when I add/change a type annotation.

**My main goal?** I would like to use the very rapid reanalysis feedback of type correctness when I change the type annotations. My questions are:

1. Does the LSP approach where type annotations are added also update the in-memory AST and thus retrigger the analysis after a textDocument/didChange command is send? Or do I need to use the pyright-internal package for that?

2. Can I add type annotations nodes directly to the in-memory AST? If so, I suppose this is then only possible with the pyright-internal package, right? I guess the other option would be to add type annotations in the Python files that are edited via my own language client and for which a LSP textDocument/didChange command is then send (i.e. the approach of question 1).

3. When I want to retrieve the type annotation slots of a function, for example. Would I need to create a specific ParseTreeWalker that retrieves the start and end range for the type annotation slots? (Because I won't be able to add type annotations via LSP when I don't know the locations of the slots.)
	a. And doesn't the range shift over for type slot 2, which is on the same line as type slot 1, when I would add a type annotation in type slot 1?

Since the tool that I'm creating is experimental, I don't really mind that the internal interface isn't as stable and could change. I just would like to know a way where I can add type annotations to my Python code and get the Pyright feedback very quickly without having to re-open files after changes as that is a costly operation. For this, it seems like that updating the in-memory AST exactly provides this.

Looking forward to your answer!

**Reply by erictraut** [reply]
Yes, pyright internally parses the text and creates a parse tree. This is an internal data structure that is never exposed via the LSP interface. The LSP interface operates on text content.

When the text content for a source file changes, pyright internally discards the old parse tree and any other data structures that were built using the old parse tree (including binder information and types cached by the type evaluator). It recomputes these data structures on demand as needed, for example in response to an LSP request.

Pyright isn't designed to allow mutation of the parse tree, so adding or removing nodes to/from the parse tree would break many internal assumptions for all of the layers that rely on the parse tree (including the binder, type evaluator, etc.). You'd need to throw out all of the stale data structures that were built from the older parse tree, which means you've gained almost no performance over simply recreating a new parse tree from mutated text. I recommend that you just mutate the text and allow pyright to operate on the modified text contents, as it was designed to do. Tokenization and parsing of text is fast relative to type evaluation. You can get a sense for the amount of time required for each phase of analysis by using the `--stats` command-line option. For example, here's a typical breakdown:

```
Found 1035 source files
pyright 1.1.325
0 errors, 0 warnings, 0 informations 
Completed in 12.982sec

Analysis stats
Total files parsed and bound: 2251
Total files checked: 1035

Timing stats
Find Source Files:    0.19sec
Read Source Files:    0.64sec
Tokenize:             0.34sec
Parse:                0.54sec
Resolve Imports:      1.15sec
Bind:                 0.6sec
Check:                8.7sec
Detect Cycles:        0.15sec
```

You can see that tokenization + parsing required on average 0.39ms per file. The tokenization + parsing time will be linearly proportional to the length of the file in characters, so longer files will take more time.

Since your tool is experimental, I presume that performance optimization isn't a top priority anyway.

One thought is that your client could build and maintain its own parse tree, which it mutates. It would then output text based on the parse tree and send the resulting text to the pyright LSP interface. To do this, you'd want a CST (concrete syntax tree) rather than an AST, and it would need to record some details that the pyright parse tree doesn't record (such as indent amounts) so it can properly recreate the original text from the parse tree. Depending on the language you're writing your tool in, you may be able to find an off-the-shelf library that does this.

**Reply by MarkBekooy** [reply]
Thanks once again! Indeed maintaining a CST (concrete syntax tree) is currently the way to go I believe. For this, I'm going to use LibCST which also forces me to create the client in Python.
My client starts the Pyright language server on stdio with `pyright-langserver --stdio`.

This is currently my **Python code**:
<details>
<summary>Click to expand code</summary>

```python
import os
import subprocess
import threading

import pylspclient

class ReadPipe(threading.Thread):
    def __init__(self, pipe):
        threading.Thread.__init__(self)
        self.pipe = pipe

    def run(self):
        line = self.pipe.readline().decode("utf-8")
        while line:
            print(line)
            line = self.pipe.readline().decode("utf-8")

def main():
    p = subprocess.Popen(
        args=["pyright-langserver", "--stdio"],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )

    read_pipe = ReadPipe(p.stderr)
    read_pipe.start()
    json_rpc_endpoint = pylspclient.JsonRpcEndpoint(p.stdin, p.stdout)
    lsp_endpoint = pylspclient.LspEndpoint(json_rpc_endpoint)

    lsp_client = pylspclient.LspClient(lsp_endpoint)

    capabilities = {
        "textDocument": {
            "synchronization": {"dynamicRegistration": True},
            "publishDiagnostics": {"relatedInformation": True},
        },
        "workspace": {
            "applyEdit": True,
            "workspaceEdit": {
                "documentChanges": True,
            },
            "workspaceFolders": True,
        },
    }

    cwd = os.getcwd()
    root_uri = f"file://{cwd}/"

    workspace_folders = [{"name": "python-lsp", "uri": root_uri}]

    lsp_client.initialize(
        processId=p.pid,
        rootPath=None,
        rootUri=root_uri,
        initializationOptions=None,
        capabilities=capabilities,
        trace="verbose",
        workspaceFolders=workspace_folders,
    )
    lsp_client.initialized()

    file_path = "<path>/lsp-mark-python/src/example/example.py"
    uri = "file://" + file_path
    text = open(file_path, "r").read()
    languageId = pylspclient.lsp_structs.LANGUAGE_IDENTIFIER.PYTHON
    version = 1
    lsp_client.didOpen(
        pylspclient.lsp_structs.TextDocumentItem(uri, languageId, version, text)
    )

    file_path_changed = "<path>/lsp-mark-python/src/example/example-changed.py"
    new_text = open(file_path_changed, "r").read()
    document = pylspclient.lsp_structs.VersionedTextDocumentIdentifier(uri, version)
    change = pylspclient.lsp_structs.TextDocumentContentChangeEvent(new_text)
    lsp_client.didChange(
        pylspclient.lsp_structs.DidChangeTextDocumentParams(document, [change])
    )

    lsp_client.shutdown()
    lsp_client.exit()


if __name__ == "__main__":
    main()
```
</details>

And that results in the following responses in the terminal:
```
SENDING: {"jsonrpc": "2.0", "id": 0, "method": "initialize", "params": {"processId": 84716, "rootPath": null, "rootUri": "file://<path>\\lsp-mark-python/", "initializationOptions": null, "capabilities": {"textDocument": {"synchronization": {"dynamicRegistration": true}, "publishDiagnostics": {"relatedInformation": true}}, "workspace": {"applyEdit": true, "workspaceEdit": {"documentChanges": true}, "workspaceFolders": true}}, "trace": "verbose", "workspaceFolders": [{"name": "python-lsp", "uri": "file://<path>\\lsp-mark-python/"}]}}

RECIEVED MESSAGE: {'jsonrpc': '2.0', 'method': 'window/logMessage', 'params': {'type': 3, 'message': 'Pyright language server 1.1.316 starting'}}
{'jsonrpc': '2.0', 'method': 'window/logMessage', 'params': {'type': 3, 'message': 'Pyright language server 1.1.316 starting'}}

RECIEVED MESSAGE: {'jsonrpc': '2.0', 'method': 'window/logMessage', 'params': {'type': 3, 'message': 'Server root directory: C:\\Users\\markb\\.cache\\pyright-python\\1.1.316\\node_modules\\pyright/dist/'}}
{'jsonrpc': '2.0', 'method': 'window/logMessage', 'params': {'type': 3, 'message': 'Server root directory: C:\\Users\\markb\\.cache\\pyright-python\\1.1.316\\node_modules\\pyright/dist/'}}

RECIEVED MESSAGE: {'jsonrpc': '2.0', 'method': 'window/logMessage', 'params': {'type': 3, 'message': 'Starting service instance "python-lsp"'}}
{'jsonrpc': '2.0', 'method': 'window/logMessage', 'params': {'type': 3, 'message': 'Starting service instance "python-lsp"'}}

RECIEVED MESSAGE: {'jsonrpc': '2.0', 'id': 0, 'result': {'capabilities': {'textDocumentSync': 2, 'definitionProvider': {'workDoneProgress': True}, 'declarationProvider': {'workDoneProgress': True}, 'typeDefinitionProvider': {'workDoneProgress': True}, 'referencesProvider': {'workDoneProgress': True}, 'documentSymbolProvider': {'workDoneProgress': True}, 'workspaceSymbolProvider': {'workDoneProgress': True}, 'hoverProvider': {'workDoneProgress': True}, 'documentHighlightProvider': {'workDoneProgress': True}, 'renameProvider': {'prepareProvider': True, 'workDoneProgress': True}, 'completionProvider': {'triggerCharacters': ['.', '[', '"', "'"], 'resolveProvider': True, 'workDoneProgress': True, 'completionItem': {'labelDetailsSupport': True}}, 'signatureHelpProvider': {'triggerCharacters': ['(', ',', ')'], 'workDoneProgress': True}, 'codeActionProvider': {'codeActionKinds': ['quickfix', 'source.organizeImports'], 'workDoneProgress': True}, 'executeCommandProvider': {'commands': [], 'workDoneProgress': True}, 'callHierarchyProvider': True, 'workspace': {'workspaceFolders': {'supported': True, 'changeNotifications': True}}}}}

SENDING: {"jsonrpc": "2.0", "method": "initialized", "params": {}}

SENDING: {"jsonrpc": "2.0", "method": "textDocument/didOpen", "params": {"textDocument": {"uri": "file://<path>/lsp-mark-python/src/example/example.py", "languageId": "python", "version": 1, "text": "def multiply(a: int, b: int):\n    return a * b\n\n\ndef sum(a: int, b: int):\n    return a + b\n"}}}

SENDING: {"jsonrpc": "2.0", "method": "textDocument/didChange", "params": {"textDocument": {"uri": "file://<path>/lsp-mark-python/src/example/example.py", "version": 1}, "contentChanges": [{"text": "def multiply(a: int, b: int) -> str:\n    return a * b\n\n\ndef sum(a: int, b: int):\n    return a + b\n"}]}}
```
**Now, I have some questions:**
I expected this to trigger the `textDocument/pubilishDiagnostics` response since I'm sending the `textDocument/didChange` notification and the multiply function now has an incorrect `str` return type. However, I don't get anything printed in the console...

So, do I need some extra code that listens for this event? I'm already printing all responses to the console, so it seems like it didn't get triggered in the first place.

Furthermore, is there an easy way to debug the responses from the Pyright language server? (i.e. see the communication messages back and forth). I've tried https://microsoft.github.io/pyright/#/build-debug?id=debugging-pyright, on the cloned Pyright repo, but that didn't work because I get:
```
Pyright has detected that the Pylance extension is installed. Pylance includes the functionality of Pyright, and running both of these extensions can lead to problems. Pyright will disable itself. Uninstall or disable Pyright to avoid this message.
```
But I don't expect that uninstalling Pylance in this newly opened VS Code instance is the way to go. Attaching the debugger is logged in the VS Code output channel, but no breakpoints get hit.

I hope you can help me out with these 2 questions and thanks already!

**Reply by erictraut** [reply]
You'll need to uninstall (or at least disable) Pylance in the debug instance of VS Code if you want pyright to run. Or modify the pyright code to disable the check for pylance. Once you get debugging working, it should be straightforward to set breakpoints in the pyright code and further diagnose where the LSP communications are going awry.
